include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -183,76 +183,86 @@ struct __init_processing
+{
     _Operation1 __binary_op;
     _Operation2 __unary_op;
 
     template <typename _Size, typename... _Acc>
     _Tp
-    seq(const _Size __n, const _Acc&... __acc) const
+    operator()(const _Size __n, const _Acc&... __acc) const
     {
-        return apply_unary_binary(__n, 0, __binary_op, __unary_op, __acc...);
+        _Tp __result = __unary_op(0, __acc...);
+        // Add neighbour to the current __result
+        for (_Size __i = 1; __i < __n; ++__i)
+            __result = __binary_op(__result, __unary_op(__i, __acc...));
+        return __result;
     }
+};
+
+template <typename _ExecutionPolicy, ::std::size_t __iters_per_work_item, typename _Operation1, typename _Operation2>
+struct transform_reduce_known
+{
+    _Operation1 __binary_op;
+    _Operation2 __unary_op;
 
-    // This routine will do nothing if the work item operates outside the bounds of __n. However, using checks
-    // (__global_idx * __iters_per_work_item) < __n) within the calling routine instead degrades the performance.
     template <typename _Size, typename _AccLocal, typename... _Acc>
     void
-    parallel(const ::std::uint16_t __local_id, const _Size __n, const ::std::size_t __dynamic_iters_per_work_item,
-             const ::std::size_t __global_id, const ::std::size_t __global_offset, _AccLocal& __local_mem,
-             const _Acc&... __acc) const
-    {
-        ::std::size_t __adjusted_global_id = __global_offset;
-        ::std::size_t __items_to_process = __n;
-        if constexpr (__iters_per_work_item > 0)
+    operator()(const ::std::uint16_t __local_id, const _Size __n,
+               const ::std::size_t /* unused __iters_per_work_item */, const ::std::size_t __global_id,
+               const ::std::size_t __global_offset, _AccLocal& __local_mem, const _Acc&... __acc) const
+    {
+        const ::std::size_t __adjusted_global_id = __global_offset + __iters_per_work_item * __global_id;
+        const ::std::size_t __items_to_process = __n - (__iters_per_work_item * __global_id);
+        // Add neighbour to the current __local_mem
+        if (__items_to_process >= __iters_per_work_item)
         {
-            __adjusted_global_id += (__iters_per_work_item * __global_id);
-            __items_to_process -= (__iters_per_work_item * __global_id);
-            if (__items_to_process >= __iters_per_work_item)
-            {
-                __local_mem[__local_id] =
-                    apply_unary_binary(__iters_per_work_item, __adjusted_global_id, __binary_op, __unary_op, __acc...);
-            }
-            else if (__items_to_process > 0)
-            {
-                __local_mem[__local_id] =
-                    apply_unary_binary(__items_to_process, __adjusted_global_id, __binary_op, __unary_op, __acc...);
-            }
+            // Keep these statements in the same scope to allow for better memory alignment
+            typename _AccLocal::value_type __res = __unary_op(__adjusted_global_id, __acc...);
+#pragma unroll
```

Comment:
#783 provides a wrapped version of `#pragma unroll` to avoid warnings for compilers which may not support it.  
However we merge these two branches, we should end up with that wrapped version.

Reply:
Great, I will use the wrapper.